pkgs/top-level/default.nix

```diff
@@ -57,24 +56,36 @@ let
   platform = if platform_ != null then platform_
     else config.platform or platformAuto;
 
-  topLevelArguments = {
-    inherit system bootStdenv noSysDirs config crossSystem platform lib;
-  };
+  # A few packages make a new package set to draw their dependencies from.
+  # (Currently to get a cross tool chain, or forced-i686 package.) Rather than
+  # give `all-packages.nix` all the arguments to this function, even ones that
+  # don't concern it, we give it this function to "re-call" nixpkgs, inheriting
+  # whatever arguments it doesn't explicitly provide. This way,
+  # `all-packages.nix` doesn't know more than it needs too.
+  #
+  # It's OK that `args` doesn't include the defaults: they'll be
+  # deterministically inferred the same way.
+  mkPackages = newArgs: import ./. (args // newArgs);
 
   stdenvAdapters = self: super:
     let res = import ../stdenv/adapters.nix self; in res // {
       stdenvAdapters = res;
     };
 
   trivialBuilders = self: super:
-    (import ../build-support/trivial-builders.nix {
+    import ../build-support/trivial-builders.nix {
       inherit lib; inherit (self) stdenv stdenvNoCC; inherit (self.xorg) lndir;
-    });
+    };
 
-  stdenvDefault = self: super: (import ./stdenv.nix topLevelArguments) {} pkgs;
```

Comment:
@DavidEGrayson in my (current out of date) rebase of @vcunat's work, in that same phase I also build the cross tools https://github.com/Ericson2314/nixpkgs/commit/ee2a4842175f31aeef2c386a53ac65f7c8f7c3b7#diff-1640a202e3ce4fd5a204d6c39e252787R110